_posts/2019-08-09-pairwise-distance.md

```diff
@@ -0,0 +1,262 @@
+this. Though one reasonably approachable way is to use Numba. We could write
+the following to handle our in-memory case.
+
+```python
+x_cuda = cuda.to_device(x)
+```
+
+We can also handle the Dask case pretty easily by simply writing this.
+
+```python
+x_dask_cuda = da.map_blocks(cuda.to_device, x)
+```
+
+This may take a moment as data moves from the CPU to the GPU.
+
+At a first pass, we might write something Numba's CUDA JIT. Here we might need
+to tune how many threads we use per block.
+
+
+TODO Add this code
+
+
+Alternatively we could be a bit clever and write this code using `forall`. This is pretty handy as we now no longer need to think about CUDA threads per block. There is a caveat though; we need to collapse our `for`-loops into 1 loop. One can do this with a bit of ingenuity.
+
+```python
+import math
+
+
+@cuda.jit(device=True)
+def square_coords_cuda(pair_index, n):
+    pair_index = np.float32(pair_index)
+    n = np.float32(n)
+    j = (((2 * n) - 1) - math.sqrt((1 - (2 * n)) ** 2 - (8 * pair_index))) // 2
+    k = pair_index - (j * ((2 * n) - j - 1) / 2) + j + 1
+    j = np.int64(j)
+    k = np.int64(k)
+    return j, k
+
+
+@cuda.jit
+def kernel_cityblock_cuda(x, out):
+    m = x.shape[0]
+    n = x.shape[1]
+    n_pairs = (n * (n - 1)) // 2
+    pair_index = cuda.grid(1)
+    if pair_index < n_pairs:
+        # Unpack the pair index to column indices.
+        j, k = square_coords_cuda(pair_index, n)
+        # Iterate over rows, accumulating distance.
+        d = np.float32(0)
+        for i in range(m):
+            u = np.float32(x[i, j])
+            v = np.float32(x[i, k])
+            d += math.fabs(u - v)
+        # Store distance result.
+        out[pair_index] = d
+
+        
+def pairwise_cityblock_cuda(x):
+
+    # Set up output array.
+    n = x.shape[1]
+    n_pairs = (n * (n - 1)) // 2
+    out = cuda.device_array(n_pairs, dtype=np.float32)
+
+    # Let numba decide number of threads and blocks.
+    kernel_spec = kernel_cityblock_cuda.forall(n_pairs)
+    kernel_spec(x, out)
+
+    # Reshape to allow for map blocks.
+    out = out.reshape((1, out.shape[0]))
+    
+    return out
+```
+
+As with any JIT, we benefit by doing a warmup run first. Though that's easy
+enough to do. Then we are ready to compute our result.
+
+```python
+dist_cuda = pairwise_cityblock_cuda(x_cuda)
+cuda.synchronize()
+```
+
+If work needs to be done on the CPU afterwards, we can copy it back like so.
+
+```python
+dist_cuda.copy_to_host()
+```
+
+
+## Larger dataset
+
+
+If we have a few GPUs at our disposal either in a desktop or a cluster, we
+could launch a Dask cluster on it.
+
+```python
+from dask_cuda import LocalCUDACluster
+from dask.distributed import Client
+
+cluster = LocalCUDACluster()
+client = Client(cluster)
+```
+
+Then distribute our computation over that cluster.
+
+```
+dist_big_cuda = pairwise_cityblock_dask(x_big_dask_cuda, f=pairwise_cityblock_cuda).compute()
+```
```

Comment:
It'd probably be worth making this into a notebook so that it can be downloaded and executed, it'd also shake out any bugs in the code blocks in the document? Having some canned/expected output would definitely help with ensuring that algorithm translations etc are done correctly.

Reply:
Yeah this started as a notebook originally. Then we moved it into this blogpost form so we could publish and share it a bit more broadly.